crates/core/tedge_agent/src/entity_manager/server.rs

```diff
@@ -307,23 +315,39 @@ impl EntityStoreServer {
     async fn deregister_entity(&mut self, topic_id: &EntityTopicId) -> Vec<EntityMetadata> {
         let deleted = self.entity_store.deregister_entity(topic_id);
         for entity in deleted.iter().rev() {
-            for twin_key in entity.twin_data.keys() {
-                let topic = self.mqtt_schema.topic_for(
-                    &entity.topic_id,
-                    &Channel::EntityTwinData {
-                        fragment_key: twin_key.to_string(),
-                    },
-                );
-                let clear_twin_msg = MqttMessage::new(&topic, "").with_retain();
-
-                self.publish_message(clear_twin_msg).await;
+            let mqtt_config = self
+                .mqtt_config
+                .clone()
+                .with_no_session()
+                .with_subscriptions(self.mqtt_schema.topics(
+                    EntityFilter::Entity(&entity.topic_id),
+                    ChannelFilter::AnyData,
+                ));
+
+            // Clear all retained metadata/data associated with the entity
+            match mqtt_channel::Connection::new(&mqtt_config).await {
```

Comment:
Ideally the connection shouldn't be established directly from here, but the mqtt actor should be enhanced to support new dynamic **connection** requests as well. But taking this simpler approach for now to avoid any conflict with #3661 by adding a new MQTT input variant (`MqttInput::Connect`) as that PR is already adding support for dynamic **subscription** requests (`MqttInput::SubscribeRequest`) in the mqtt actor.